File: Data Structures & Algorithms/search-2d-matrix/submission-7.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {

        // matrix = [[1,2,4,8],
        //           [10,11,12,13],
        //           [14,20,30,40]], target = 10

        int row = matrix.size();
        int col = matrix[0].size();
        int num_ele = row * col;

        int left = 0;
        int right = num_ele - 1;

        while (left<=right)
        {
            int mid = left + (right-left) / 2;
            int loc_row = mid / col;
            int loc_col = mid % col;

            if (matrix[loc_row][loc_col] == target) return true;

            else if (matrix[loc_row][loc_col] > target) right = mid - 1;

            else if (matrix[loc_row][loc_col] < target) left = mid + 1;
        }

        return false;
        
    }
};

File: Data Structures & Algorithms/buy-and-sell-crypto/submission-7.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:

        # prices = [10,1,5,6,7,1]

        left = 0
        right = 1
        max_profit = 0

        while right < len(prices):

            if prices[right] < prices[left]:
                left = right
            elif prices[right] > prices[left]:
                curr_profit = prices[right] - prices[left]
                max_profit = max(curr_profit, max_profit)
            else:
                pass
            right+= 1

        return max_profit

        